src/effects/VST3/VST3Effect.cpp

```diff
@@ -0,0 +1,787 @@
+}
+
+bool VST3Effect::LoadUserPreset(const RegistryPath& name)
+{
+   //Dummy
+   return true;
+}
+
+bool VST3Effect::SaveUserPreset(const RegistryPath& name)
+{
+   //Dummy
+   return true;
+}
+
+RegistryPaths VST3Effect::GetFactoryPresets()
+{
+   return { };
+}
+
+bool VST3Effect::LoadFactoryPreset(int id)
+{
+   return true;
+}
+
+bool VST3Effect::LoadFactoryDefaults()
+{
+   using namespace Steinberg;
+
+   if(mEditController != nullptr)
+   {
+      for(int i = 0, count = mEditController->getParameterCount(); i < count; ++i)
+      {
+         Vst::ParameterInfo parameterInfo;
+         if(mEditController->getParameterInfo(i, parameterInfo) == kResultOk)
+         {
+            if(parameterInfo.flags & Vst::ParameterInfo::kIsReadOnly)
+               continue;
+
+            mComponentHandler->performEdit(parameterInfo.id, parameterInfo.defaultNormalizedValue);
+            mEditController->setParamNormalized(parameterInfo.id, parameterInfo.defaultNormalizedValue);
+         }
+      }
+   }
+   return true;
+}
+
+namespace
+{
+   unsigned CountChannels(Steinberg::Vst::IComponent* component,
+      const Steinberg::Vst::MediaTypes mediaType, 
+      const Steinberg::Vst::BusDirection busDirection,
+      const Steinberg::Vst::BusType busType)
+   {
+      using namespace Steinberg;
+
+      unsigned channelsCount{0};
+
+      const auto busCount = component->getBusCount(mediaType, busDirection);
+      for(auto i = 0; i < busCount; ++i)
+      {
+         Vst::BusInfo busInfo;
+         if(component->getBusInfo(mediaType, busDirection, i, busInfo) == kResultOk)
+         {
+            if(busInfo.busType == busType)
+               channelsCount += busInfo.channelCount;
+         }
+      }
+      return channelsCount;
+   }
+}
+
+unsigned VST3Effect::GetAudioInCount()
+{
+   return CountChannels(
+      mEffectComponent,
+      Steinberg::Vst::kAudio,
+      Steinberg::Vst::kInput,
+      Steinberg::Vst::kMain);
+}
+
+unsigned VST3Effect::GetAudioOutCount()
+{
+   return CountChannels(
+      mEffectComponent,
+      Steinberg::Vst::kAudio,
+      Steinberg::Vst::kOutput,
+      Steinberg::Vst::kMain);
+}
+
+int VST3Effect::GetMidiInCount()
+{
+   //Dummy
+   return 0;
+}
+
+int VST3Effect::GetMidiOutCount()
+{
+   //Dummy
+   return 0;
+}
+
+void VST3Effect::SetSampleRate(double rate)
+{
+   mSetup.sampleRate = rate;
+}
+
+size_t VST3Effect::SetBlockSize(size_t maxBlockSize)
+{
+   //Dummy
+   mSetup.maxSamplesPerBlock = static_cast<Steinberg::int32>(maxBlockSize);
+   return mSetup.maxSamplesPerBlock;
+}
+
+size_t VST3Effect::GetBlockSize() const
+{
+   return mSetup.maxSamplesPerBlock;
+}
+
+sampleCount VST3Effect::GetLatency()
+{
+   using namespace Steinberg;
+
+   auto latency { 0u };
+   
+   for(auto& realtimeProcessor : mRealtimeGroupProcessors)
+   {
+      Vst::IAudioProcessor* audioProcessor;
+      if(realtimeProcessor->mEffectComponent->queryInterface(Vst::IAudioProcessor::iid, (void**)&audioProcessor) == kResultOk)
```

Comment:
Use `reinterpret_cast` consistently here

Reply:
And perhaps make a little inline function to do this query.  It happens in seven places.